extensions/amp-timeago/0.2/timeago.js

```diff
@@ -0,0 +1,70 @@
+/**
+ * Copyright 2019 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {createElement, useEffect, useRef, useState} from '../../../src/preact';
+import {timeago} from '../../../third_party/timeagojs/timeago';
+import {useResourcesNotify} from '../../../src/preact/utils';
+
+/**
+ * @param {!JsonObject} props
+ * @return {Preact.Renderable}
+ */
+export function Timeago(props) {
+  const [, setIsIntersecting] = useState(null);
+  const ref = useRef(null);
+
+  useEffect(() => {
+    const node = ref.current;
+    const observer = new IntersectionObserver(entries => {
+      const last = entries[entries.length - 1];
+      setIsIntersecting(last.isIntersecting);
+    });
+    if (node) {
+      observer.observe(node);
+    }
+    return () => observer.disconnect();
+  }, []);
+
+  useResourcesNotify();
+  return createElement(
+    'time',
+    {datetime: props['datetime'], ref},
+    getFuzzyTimestampValue(props)
```

Comment:
This changes the timestamp on each render. And per spec it should only do so when the element returns back to the viewport, right? Why not just do it in the `if (isIntersecting) {}` and store the value in the state instead of `isIntersecting`?

Reply:
Won't this approach will cost us rendering a fresh timestamp on `prop` mutation i.e. from amp-bind?

Reply:
- you want to update ts on either prop change onEnterViewport
- you do not want to update ts on random parent rerender or onExitViewport

Does this work?
```
export function Timeago(props) {
  const [intersectionCount, setIntersectionCount] = useState(0);
  const [timestamp, setTimestamp] = useState(null);
  const ref = useRef(null);

  // IO is used just to increment a count in state each time it enters viewport.
  useEffect(() => {
    const node = ref.current;
    const observer = new IntersectionObserver(entries => {
      const last = entries[entries.length - 1];
      if (last.isIntersecting) {
        setIntersectionCount(intersectionCount+1); 
      }
    });
    if (node) {
      observer.observe(node);
    }
    return () => observer.disconnect();
  }, []);

  // If either IO count or props change, then update timestamp.
  useEffect(() => {
    setTimestamp(getFuzzyTimestampValue(props)); 
  }, [props.datetime, intersectionCount]);

  useResourcesNotify();
  return createElement('time', {datetime: props['datetime'], ref}, timestamp);
}
```

Reply:
But, I think the code was fine as is, it doesn't matter too much if timeago renders when exiting viewport.

Reply:
Rerunning effect on prop change is very common and no problem with it that I can think of. The "this pattern is shady" in the links, I believe, is with passing arrays around.

Reply:
Also, the initial problem wasn't with the "renders when exiting viewport", but with the content changing on every external rerender with state/props being invariant.